server/modules/instantiation/port-provider.js

```diff
@@ -52,28 +57,31 @@ module.exports = function () {
         }
 
         providePort(expectedPortRanges) {
-            var i, port;
-            for (i = 0; i < expectedPortRanges.length; i += 1) {
-                for (port = expectedPortRanges[i].minPort; port <= expectedPortRanges[i].maxPort; port += 1) {
-                    if (this.availablePortRanges.contains(port) && !this.externalPortRanges.contains(port)) {
-                        this.addExternalPortRange(port, port);
-
-                        return port;
-                    }
-                }
-            }
-
-            throw new Error('Failed to allocate port.');
+            return Promise
+                .any(expectedPortRanges.map(portRange => this.findPort(portRange)))
+                .catch(_ => new Error('Failed to allocate port.'));
         }
-    }
 
-    var portProvider = new PortProvider();
+        findPort(portRange) {
+            portfinder.basePort = portRange.minPort;
+
+            return new Promise((resolve, reject) => {
+                portfinder.getPortPromise()
+                    .then(port => {
+                        if (this.availablePortRanges.contains(port) && portRange.contains(port)) {
+                            resolve(port);
+
+                            return;
```

Comment:
this was another big struggle here, apparently `resolve()` doesn't terminate the execution so actually it is needed